### PR TITLE
[Feature] Support LoRA

### PR DESCRIPTION
**Pull request by fanqiNO1**
## Motivation

Support LoRA.

## Modification

Support LoRA.

## Use cases

```python
model = dict(
    type='ImageClassifier',
    backbone=dict(
        type='LoRAModel',
        module=dict(
            type='VisionTransformer',
            arch='b',
            img_size=384,
            patch_size=16,
            drop_rate=0.1,
            init_cfg=dict(type='Pretrained', checkpoint=''), prefix='backbone'),
        alpha=16,
        rank=16,
        drop_rate=0.1,
        targets=[dict(type='qkv')]),
    neck=None,
    head=dict(
        type='VisionTransformerClsHead',
        num_classes=1000,
        in_channels=768,
        loss=dict(
            type='LabelSmoothLoss', label_smooth_val=0.1,
            mode='classy_vision'),
        init_cfg=[dict(type='TruncNormal', layer='Linear', std=2e-5)],
    ))
```

## Experiments

1. > Use the vit-p16-224 pretrained on the imagenet21k and finetune on the imagenet1k with 384px.
    > The GPU memory consumption has been reduced  from 24GB to 17GB.
    > The number of trainable parameters has been reduced from 88M to 1.2M.
    > The accuracy is `accuracy/top1`: 84.1000, `accuracy/top5`: 97.1200.
    > The `accuracy/top1` from the original paper [An Image is Worth 16x16 Words: Transformers for Image Recognition at Scale](https://arxiv.org/pdf/2010.11929.pdf) is 83.97.

2. > Use the dinov2-small and finetune on the imagenet1k.
    > The accuracy is `accuracy/top1`: 81.4360  `accuracy/top5`: 95.9140.
    > The `accuracy/top1` from the original paper [DINOv2: Learning Robust Visual Features without Supervision](https://arxiv.org/pdf/2304.07193.pdf) is 81.1 (Linear evaluation on ImageNet-1k of frozen pretrained features).

3. > Use the BLIP2 and finetune on the COCOCaption.
    > The GPU memory consumption has been reduced  from OOM (>80GB) to 60GB.
    > The result is `BLEU@4`: 42.65, `CIDEr`: 143.84.
    > The result from the original paper [BLIP-2: Bootstrapping Language-Image Pre-training with Frozen Image Encoders and Large Language Models](https://arxiv.org/pdf/2301.12597.pdf) is `BLEU@4`: 43.7, `CIDEr`: 145.8. (The original paper finetunes the whole vision backbone, but I just finetune the attention layers of the vision backbone.)

## Checklist

**Before PR**:

- [ ] Pre-commit or other linting tools are used to fix the potential lint issues.
- [ ] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects, like MMDet or MMSeg.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

**Review by fangyixiao18 — CHANGES_REQUESTED** [review]
1. provide an example lora config of ViT
2. add a lora weights merge script for users

**Review by fangyixiao18 — APPROVED** [review]

File: tools/model_converters/merge_lora_weight.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    prefix_split = prefix.split('.')
+    for i in prefix_split:
+        model = getattr(model, i)
+    return model.scaling
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(description='Merge LoRA weight')
+    parser.add_argument('cfg', help='cfg path')
+    parser.add_argument('src', help='src detectron lora model path')
```
Comment (mzr1996) [inline tools/model_converters/merge_lora_weight.py:77]: `detectron lora model`?

File: tests/test_models/test_peft/test_lora.py [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+    incompatible_keys = lora_model.load_state_dict(state_dict, strict=True)
+    assert str(incompatible_keys) == '<All keys matched successfully>'
+
+
+# @pytest.mark.skipif(
+#     digit_version(TORCH_VERSION) < digit_version('1.9.0'),
+#     reason='get_submodule requires torch >= 1.9.0')
+# def test_lora_huggingface():
+#     module = dict(type='OPTForCausalLM', name_or_path='facebook/opt-125m')
+
```
Comment (mzr1996) [inline tests/test_models/test_peft/test_lora.py:130]: Remove it if it's useless

**Review by mzr1996 — CHANGES_REQUESTED** [review]

**Review by mzr1996 — APPROVED** [review]
